class.jetpack.php

```diff
@@ -738,11 +732,65 @@ function( $callable ) {
 		add_filter( 'jetpack_token_request_body', array( __CLASS__, 'filter_token_request_body' ) );
 	}
 
+	/**
+	 * Runs on plugins_loaded but with a higher priority to make sure we're initializing
+	 * before other hooks.
+	 *
+	 * @action plugins_loaded
+	 * @priority 5
+	 */
+	public function early_initalization() {
+		/*
+		 * Enable enhanced handling of previewing sites in Calypso
+		 */
+		if ( self::is_active() ) {
+			require_once JETPACK__PLUGIN_DIR . '_inc/lib/class.jetpack-iframe-embed.php';
+			add_action( 'init', array( 'Jetpack_Iframe_Embed', 'init' ), 9, 0 );
+			require_once JETPACK__PLUGIN_DIR . '_inc/lib/class.jetpack-keyring-service-helper.php';
+			add_action( 'init', array( 'Jetpack_Keyring_Service_Helper', 'init' ), 9, 0 );
+		}
+
+		Sync_Main::init();
+
+		// Check for WooCommerce support.
+		Sync_Actions::initialize_woocommerce();
+
+		// Check for WP Super Cache.
+		Sync_Actions::initialize_wp_super_cache();
+	}
+
+	/**
+	 * Runs on plugins_loaded but with a lower priority.
+	 *
+	 * @action plugins_loaded
+	 * @priority 90
+	 */
+	public function late_initalization() {
+		/*
+		 * Init after plugins loaded and before the `init` action. This helps with issues where plugins init
+		 * with a high priority or sites that use alternate cron.
+		 */
+		Sync_Actions::init();
+
+		self::plugin_textdomain();
+		self::load_modules();
+	}
+
 	/**
 	 * Runs after all the plugins have loaded but before init.
 	 */
 	function after_plugins_loaded() {
 
+		/**
+		 * Fires when Jetpack is fully loaded and ready. This is the point where it's safe
+		 * to instantiate classes from packages and namespaces that are managed by the Jetpack Autoloader.
+		 *
+		 * @since 8.1.0
+		 *
+		 * @param Jetpack $jetpack the main plugin class object.
+		 */
+		do_action( 'jetpack_loaded', $this );
```

Comment:
So we're good with exposing the entire Jetpack object to the public?

I realize that to be pluggable this needs to happen, but just wanted to confirm.

Reply:
Yes, because right now the code that needs the Jetpack object just calls `Jetpack::init`, which is way dirtier - we have no control over when anyone decides to init Jetpack, and whether it's too early in the execution cycle or not.